clash-lib/src/Clash/Normalize/Transformations.hs

```diff
@@ -424,9 +424,63 @@ caseCase _ e = return e
 {-# SCC caseCase #-}
 
 -- | Inline function with a non-representable result if it's the subject
--- of a Case-decomposition
+-- of a Case-decomposition. It's a custom topdown traversal that -for efficiency
+-- reasons- does not explore alternative of cases whose subject triggered an
+-- 'inlineNonRepWorker'.
 inlineNonRep :: HasCallStack => NormRewrite
-inlineNonRep _ e@(Case scrut altsTy alts)
+inlineNonRep _ctx = fmap snd . go
+ where
+  go :: HasCallStack => Term -> NormalizeSession (Bool, Term)
+  go e0 = inlineNonRepWorker e0 >>= \case
+    (True, e1) ->
+      return (True, e1)
+    (False, e1) ->
```

Comment:
Another thing: you return a `Bool` which indicates whether there was a change, but the `Writer` part of this monad tracks if there was a change (as far as I can tell anyway, it's rather unhelpfully marked as `Any` in the types module). However I have seen things like

```haskell
(e1, Monoid.getAny -> hasChanged) <- Writer.listen $ rw ctx e
```

in other parts of the codebase, which make me think you could instead write

```haskell
Writer.listen (inlineNonRepWorker e0) >>= \case
  (e1, Monoid.getAny -> True)  -> pure e1
  (e1, Monoid.getAny -> False) -> case e1 of ...
```

and use the fact that `return` and `changed` will determine what happens

Reply:
I've considered it but decided against it because you need to reset the flag if you're handling things like `App e1 e2` or `Letrec bindings_ body`. I.e., if `e1` triggers a rewrite that does _not_ mean `e2` shouldn't try to inline non rep constructs.

Reply:
In `inlineNonRepWorker` or `inlineNonRep.go`? If you mean `inlineNonRep.go` then I'm not sure I see what you mean, you appear to rewrite different parts of `App` and `Letrec` then just or the results at the end

Reply:
Yeah alright, I misunderstood what `listen` was doing. I like the advice.